keps/976-plain-pods/README.md

```diff
@@ -538,39 +536,51 @@ spec:
-gated to ungated.
+The Pod reconciler will run in a "composable" mode: a mode where a Workload is composed of multiple
+objects. The `jobframework.Reconciler` will be reworked to accomodate this.
 
-In the Pod event handler, we decrement the counter when we see a transition from having
-the scheduling gate `kueue.x-k8s.io/admission` to not having it.
+After a Workload is admitted, each Pod that owns the workload enters the reconciliation loop.
+The reconciliation loop collects all the Pods that are not Failed and constructs an in-memory
+Workload. If there is an existing Workload in the cache and it has smaller Pod counters than the
+in-memory Workload, then it is considered unmatching and the Workload is evicted.
 
 In the Pod-group reconciler:
 1. If the Pod is not terminated,
-  create a Workload for the pod group if one does not exist.
-2. If the Pod is terminated,
-   - If the Workloald doesn't exist or the workload is finished, remove the finalizer.
-3. ungated_pods_in_client: the number of non-terminated pods in the client that are admitted.
-   We only look at non-terminated pods to allow for terminated pods to be replaced.
-4. ungated_pods = ungated_pods_in_client + expected_ungated_pods. Note that this might temporarily
-   lead to double counting.
-2. For gated pods:
-  - If ungated_pods < admission.count, remove the gate, set nodeSelector, an increase
-    expected_ungated_pods
-  - Else,
-    - If ungated_pods_in_informer < admission.count, we can't admit this Pod now to prevent
-      overbooking, but requeue this Pod for retry.
-    - Else, remove finalizer and delete the Pod, as it's beyond the allowed admission.
-5. If the number of terminated pods with a finalizer is greater than or equal to the admission
-  count, and there are no non-terminated Pods, mark the Workload as Finished and remove the
-  finalizers from the Pods.
-
-Note that we are only removing Pod finalizers once the Workload is finished. This is a simple way
-of managing finalizers, but it might lead to too many Pods lingering in etcd for a long time after
-terminated. In a future version, we can consider a better scheme similar to [Pod tracking in Jobs](https://kubernetes.io/blog/2022/12/29/scalable-job-tracking-ga/).
+   create a Workload for the pod group if one does not exist.
+2. Remove Pod finalizers if the Pod is terminated and the Workload is finished, has a deletion
+   timestamp or is finished.
+3. Build the in-memory Workload. If its podset counters are greater than the stored Workload,
+   then evict the Workload.
+4. For gated pods:
+   - remove the gate, set nodeSelector
+5. If the number of succeeded pods is equal to the admission count, mark the Workload as Finished
+   and remove the finalizers from the Pods.
+
+Note that we are only removing Pod finalizers once the Workload is finished or if the Pods are
+Failed. This is a simple way of managing finalizers, but it might lead to too many Pods lingering
```

Comment:
```suggestion
Note that we are only removing Pod finalizers once the Workload is finished or if the Pods are
Failed and replaced. This is a simple way of managing finalizers, but it might lead to too many Pods lingering
```

Reply:
Oops, I actually wanted to say that we remove finalizers only when the workload is finished.